### PR TITLE
Implement search-data using dedicated API calls.

### PR DESCRIPTION
**Pull request by cmouse**
Provides searchRecords and searchComments API calls for backends to perform better optimized search for data. 
- [x] Remotebackend
- [x] gSQL backends
- [x] BIND backend
- [x] Tests

### CODE REVIEW

**Comment by cmouse** [comment]
@tuxis-ie you might be interested to test this.

**Comment by cmouse** [comment]
Pending on #2665

**Comment by cmouse** [comment]
Ready for review, though.

File: pdns/ws-auth.cc [hunk 1]
```diff
@@ -1079,81 +1079,88 @@ static void apiServerSearchData(HttpRequest* req, HttpResponse* resp) {
...
-#if 0
-    di.backend->list(di.zone, di.id, true); // incl. disabled
-    while(di.backend->get(rr)) {
-      if (!rr.qtype.getCode())
-        continue; // skip empty non-terminals
-
-      if (pdns_ci_find(rr.qname, q) == string::npos && pdns_ci_find(rr.content, q) == string::npos)
-        continue;
-
+  if (B.searchRecords(q, maxEnts, result_rr))
```
Comment (zeha) [inline pdns/ws-auth.cc]: it'd be a lot nicer if the if() wouldn't be needed. result_rr is an empty vector anyway, no?

File: pdns/ws-auth.cc [hunk 2]
```diff
@@ -1079,81 +1079,88 @@ static void apiServerSearchData(HttpRequest* req, HttpResponse* resp) {
...
       object.AddMember("content", jcontent, doc.GetAllocator());
       doc.PushBack(object, doc.GetAllocator());
     }
+  }
 
-    di.backend->listComments(di.id);
-    while(di.backend->getComment(comment)) {
-      if (pdns_ci_find(comment.qname, q) == string::npos && pdns_ci_find(comment.content, q) == string::npos)
-        continue;
+  if (B.searchComments(q, maxEnts, result_c))
```
Comment (zeha) [inline pdns/ws-auth.cc]: as above

File: pdns/backends/gsql/gsqlbackend.cc [hunk 3]
```diff
@@ -1501,6 +1506,105 @@ string GSQLBackend::directBackendCmd(const string &query)
...
 }
 
+bool GSQLBackend::searchRecords(const string &pattern, int maxResults, vector<DNSResourceRecord>& result)
+{
+  try {
+    string escaped_pattern = boost::replace_all_copy(pattern,"\\","\\\\");
+    boost::replace_all(escaped_pattern,"_","\\_");
+    boost::replace_all(escaped_pattern,"%","\\%");
+    boost::replace_all(escaped_pattern,"*","%");
+    boost::replace_all(escaped_pattern,"?","_");
```
Comment (zeha) [inline pdns/backends/gsql/gsqlbackend.cc]: no idea if this is SQLi safe...

File: pdns/backends/gsql/gsqlbackend.cc [hunk 4]
```diff
@@ -1501,6 +1506,105 @@ string GSQLBackend::directBackendCmd(const string &query)
...
+
+bool GSQLBackend::searchComments(const string &pattern, int maxResults, vector<Comment>& result)
+{
+  Comment c;
+  try {
+    string escaped_pattern = boost::replace_all_copy(pattern,"\\","\\\\");
+    boost::replace_all(escaped_pattern,"_","\\_");
+    boost::replace_all(escaped_pattern,"%","\\%");
+    boost::replace_all(escaped_pattern,"*","%");
+    boost::replace_all(escaped_pattern,"?","_");
```
Comment (zeha) [inline pdns/backends/gsql/gsqlbackend.cc]: move the replace_ stuff into a local helper function, for less code duplication?

File: pdns/backends/gsql/gsqlbackend.cc [hunk 5]
```diff
@@ -1501,6 +1506,105 @@ string GSQLBackend::directBackendCmd(const string &query)
...
+      if(d_dnssecQueries)
+        r.auth = !row[7].empty() && row[7][0]=='1';
+      else
+        r.auth = 1;
+
+      r.disabled = !row[5].empty() && row[5][0]=='1';
+
+      r.domain_id=atoi(row[4].c_str());
+      result.push_back(r);
+    }
```
Comment (zeha) [inline pdns/backends/gsql/gsqlbackend.cc]: move the result mangling into a helper?

File: docs/markdown/authoritative/backend-remote.md [hunk 6]
```diff
@@ -1045,6 +1045,38 @@ Content-Type: text/javascript; charset=utf-8
 {"result":"PONG"}
 ```
 
+### `searchRecords`
+Can be used to search records from the backend. This is used by web api.
+
+* Mandatory: no
+* Parameters: pattern, maxResults
+* Reply: same as (lookup)(backend-remote.md#lookup) or false to indicate failed search
```
Comment (pieterlexis) [inline docs/markdown/authoritative/backend-remote.md]: wrong markdown syntax (should be `[link text](url)`). Also, it's on the same page so the following should work: `[lookup](#lookup)`

File: pdns/ws-auth.cc (hunk 1 above)
Comment (cmouse) [inline pdns/ws-auth.cc]: false indicates that the search failed for some reason, not that it did not give results. result_rr will be unspecified in that case.

File: pdns/backends/gsql/gsqlbackend.cc (hunk 3 above)
Comment (cmouse) [inline pdns/backends/gsql/gsqlbackend.cc]: no idea what is SQLi

**Comment by cmouse** [comment]
Nits fixed

**Comment by cmouse** [comment]
Ready for merge.

**Pull request merged**